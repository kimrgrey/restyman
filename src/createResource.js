import { config } from './config'
import { defineMethods } from './methods'
import { createEndpoint } from './createEndpoint'
import { NAMESPACES, forOwn } from './utils'

export const createResource = (parameters) => {
  const resources = {}
  const memberEndpoints = {}

  const collection = function (code, endpoint = createEndpoint()) {
    this[code] = _assignEndpoint(endpoint)
    return endpoint
  }

  const member = (code, endpoint = createEndpoint()) => {
    memberEndpoints[code] = endpoint
    return endpoint
  }

  const subresources = (_resources) => Object.assign(resources, _resources)

  const define = function (enhance) {
    enhance(NAMESPACES.reduce(_definition(this), {}))
  }

  const _create = (path) => {
    const API = {
      collection,
      member,
      subresources,
      define,
      getPath: () => path
    }
    return Object.assign((id) => _createResource(`${path}/${id}`), API)
  }

  const _createResource = (path) => {
    const resource = createResource({ path })
    forOwn(resources, (sub, code) => {
      resource[code] = _copyWithPrefix(sub, resource.getPath())
    })
    forOwn(memberEndpoints, (endpoint, code) => {
      resource[code] = _assignEndpoint(endpoint)
    })
    return resource
  }

  const _definition = (resource) => (result, namespace) => {
    result[namespace] = (code, request) => resource[namespace](code).request(request)
    return result
  }

  const _assignEndpoint = (endpoint) => function () {
    const req = (parameters.factory || config.factory)(this.getPath())
    return endpoint.execute(Object.assign({ req }, parameters), ...arguments)
  }

  const _copyWithPrefix = function (resource, prefix) {
    const copy = _create(`${prefix}/${resource.getPath()}`)
    forOwn(resource, (v, k) => !copy[k] && (copy[k] = v))
    return copy
  }

  return defineMethods(_create(parameters.path))
}
