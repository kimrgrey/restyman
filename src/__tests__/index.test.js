/* eslint-env jest */
import axios from 'axios'
import moxios from 'moxios'
import createResource, { setAxiosFactory } from '..'

setAxiosFactory(({ path }) => (
  axios.create({
    baseURL: `/${path}`
  })
))

describe('restyman', () => {
  let companies, users

  beforeEach(() => {
    moxios.install()

    companies = createResource({ path: 'companies' })
    users = createResource({ path: 'users' })
  })

  afterEach(() => {
    moxios.uninstall()
  })

  describe('getPath', () => {
    it('generates correct path for collection', () => {
      expect(companies.getPath()).toMatchSnapshot()
    })

    it('generates correct path for member', () => {
      expect(companies(1).getPath()).toMatchSnapshot()
    })

    it('generates correct path for subresource collection', () => {
      companies.subresources({ users })

      expect(companies(1).users.getPath()).toMatchSnapshot()
    })

    it('generates correct path for subresource member', () => {
      companies.subresources({ users })

      expect(companies(1).users(1).getPath()).toMatchSnapshot()
    })

    it('keeps correct path for initial resource', () => {
      companies.subresources({ users })

      companies(1)
      expect(users.getPath()).toMatchSnapshot()
    })
  })
})
