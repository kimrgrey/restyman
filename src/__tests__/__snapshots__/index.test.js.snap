// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`restyman getPath generates correct path for collection 1`] = `"companies"`;

exports[`restyman getPath generates correct path for member 1`] = `"companies/1"`;

exports[`restyman getPath generates correct path for subresource collection 1`] = `"companies/1/users"`;

exports[`restyman getPath generates correct path for subresource member 1`] = `"companies/1/users/1"`;

exports[`restyman getPath keeps correct path for initial resource 1`] = `"companies/1/users"`;
